NeoML/src/TraditionalML/KMeansClustering.cpp

```diff
@@ -55,9 +55,729 @@ static CPtr<CDnnBlob> createWeightBlob( IMathEngine& mathEngine, const IClusteri
+			Second( second ),
+			SecondHeight( secondHeight ),
+			Result( result )
+		{}
+
+		void CallRun( int threadIndex )
+		{
+			int firstHeightStart;
+			int firstHeightCount;
+			int secondHeightStart;
+			int secondHeightCount;
+			if( OmpGetTaskIndexAndCount2D( FirstHeight, /*alignX*/1, SecondHeight, FloatAlignment,
+				firstHeightStart, firstHeightCount, secondHeightStart, secondHeightCount,
+				ThreadCount, threadIndex ) )
+			{
+				const CConstFloatHandle& first = First + firstHeightStart * FirstWidth;
+				const CConstFloatHandle& second = Second + secondHeightStart * FirstWidth;
+				const CFloatHandle& result = Result + firstHeightStart * SecondHeight + secondHeightStart;
+
+				MathEngine.MultiplyMatrixByTransposedMatrix( first, firstHeightCount, FirstWidth, FirstWidth/*RowSize*/,
+					second, secondHeightCount, FirstWidth/*RowSize*/, result, SecondHeight/*RowSize*/, 0/*resBufferSize*/ );
+			}
+		}
+	protected:
+		const int ThreadCount;
+		IMathEngine& MathEngine;
+		const CConstFloatHandle& First;
+		const int FirstHeight;
+		const int FirstWidth;
+		const CConstFloatHandle& Second;
+		const int SecondHeight;
+		const CFloatHandle& Result;
+	};
+
+	CThreadTaskMxMT task( owner.GetThreadPool()->Size(), mathEngine,
+		first, firstHeight, firstWidth, second, secondHeight, result );
+
+	NEOML_NUM_THREADS( *owner.GetThreadPool(), &task, []( int threadIndex, void* ptr ) {
+		( ( CThreadTaskMxMT* )ptr )->CallRun( threadIndex );
+	} );
+}
+
+static void matrixDiagMultiply(
+	const CKMeansClustering& owner, IMathEngine& mathEngine,
+	const CConstFloatHandle& first, int firstSize,
+	const CConstFloatHandle& second, int secondWidth,
+	const CFloatHandle& result, int resultBufferSize )
+{
+	if( !isThreadTaskRelevant( firstSize * secondWidth ) ) {
+		mathEngine.MultiplyDiagMatrixByMatrix( first, firstSize,
+			second, secondWidth, result, resultBufferSize );
+		return;
+	}
+
+	NeoAssert( resultBufferSize >= firstSize * secondWidth );
+	struct CThreadTaskDiagMxM final {
+		CThreadTaskDiagMxM( IMathEngine& mathEngine, const CConstFloatHandle& first, int firstSize,
+				const CConstFloatHandle& second, int secondWidth, const CFloatHandle& result ) :
+			MathEngine( mathEngine ),
+			First( first ),
+			FirstSize( firstSize ),
+			Second( second ),
+			SecondWidth( secondWidth ),
+			Result( result )
+		{}
+
+		void CallRun( int /*threadIndex*/ )
+		{
+			for( int i = 0; i < FirstSize; ++i ) {
+				const CConstFloatHandle& multiplier = First + i;
+				MathEngine.VectorMultiply( Second + i * SecondWidth, Result + i * SecondWidth, SecondWidth, multiplier );
+			}
+		}
+	protected:
+		IMathEngine& MathEngine;
+		const CConstFloatHandle& First;
+		const int FirstSize;
+		const CConstFloatHandle& Second;
+		const int SecondWidth;
+		const CFloatHandle& Result;
+	};
+
+	CThreadTaskDiagMxM task( mathEngine, first, firstSize, second, secondWidth, result );
+
+	NEOML_NUM_THREADS( *owner.GetThreadPool(), &task, []( int threadIndex, void* ptr ) {
```

Comment:
Where do you split the task between threads if `CThreadTaskDiagMxM::CallRun` ignores `threadIndex`?

Reply:
fixed